source/credentials_provider_cognito.c

```diff
@@ -0,0 +1,925 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/auth/credentials.h>
+
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/common/clock.h>
+#include <aws/common/json.h>
+#include <aws/common/string.h>
+#include <aws/http/connection.h>
+#include <aws/http/connection_manager.h>
+#include <aws/http/request_response.h>
+#include <aws/http/status_code.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/retry_strategy.h>
+#include <aws/io/socket.h>
+#include <aws/io/stream.h>
+#include <aws/io/tls_channel_handler.h>
+
+#include <inttypes.h>
+
+#define COGNITO_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS 5
+#define COGNITO_MAX_RETRIES 8
+#define HTTP_REQUEST_BODY_INITIAL_SIZE 1024
+#define HTTP_RESPONSE_BODY_INITIAL_SIZE 4096
+
+static void s_on_connection_manager_shutdown(void *user_data);
+static void s_on_connection_setup_fn(struct aws_http_connection *connection, int error_code, void *user_data);
+
+struct aws_cognito_login {
+    struct aws_byte_cursor identity_provider_name;
+    struct aws_byte_cursor identity_provider_token;
+    struct aws_byte_buf login_buffer;
+};
+
+static int s_aws_cognito_login_init(
+    struct aws_cognito_login *login,
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor identity_provider_name,
+    struct aws_byte_cursor identity_provider_token) {
+    AWS_ZERO_STRUCT(*login);
+
+    size_t combined_len = 0;
+    if (aws_add_size_checked(identity_provider_name.len, identity_provider_token.len, &combined_len)) {
+        return AWS_OP_ERR;
+    }
+
+    if (aws_byte_buf_init(&login->login_buffer, allocator, combined_len)) {
+        return AWS_OP_ERR;
+    }
+
+    login->identity_provider_name = identity_provider_name;
+    login->identity_provider_token = identity_provider_token;
+
+    if (aws_byte_buf_append_and_update(&login->login_buffer, &login->identity_provider_name)) {
+        goto on_error;
+    }
+
+    if (aws_byte_buf_append_and_update(&login->login_buffer, &login->identity_provider_token)) {
+        goto on_error;
+    }
```

Comment:
There is a helper function to do exactly this: [aws_byte_buf_init_cache_and_update_cursors()](https://github.com/awslabs/aws-c-common/blob/c815cddf645ded929f9a194b6fa6dbb1c0043427/include/aws/common/byte_buf.h#L164-L172)